render/src/umi/mod.rs

```diff
@@ -131,101 +162,139 @@ impl Umi {
         }
     }
 
-    fn read_row(&mut self, line: usize) -> Block {
+    fn read_row(&mut self, line: usize) -> Result<Block, ParserError> {
         let mut current_line = self.elements[line].clone();
-        let attributes: HashMap<String, String> =
-            serde_json::from_str(&current_line.attributes).unwrap();
+        let properties: HashMap<String, String> =
+            serde_json::from_str(&current_line.properties).unwrap_or_default();
         match current_line.kind.as_str() {
             "Heading" => {
                 let heading = Heading {
                     id: current_line.id.clone(),
                     level: unimarkup_parser::elements::atomic::HeadingLevel::try_from(
-                        attributes.get("level").unwrap().as_str(),
+                        properties
+                            .get("level")
+                            .ok_or(ParserError::MissingProperty((
+                                "level".into(),
+                                current_line.position,
+                            )))?
+                            .as_str(),
                     )
-                    .unwrap(),
+                    .ok()
+                    .ok_or(ParserError::InvalidPropertyValue((
+                        "level".into(),
+                        current_line.position,
+                    )))?,
                     content: self.read_inlines(current_line.content.clone()),
-                    attributes: (attributes.get("attributes").cloned()).filter(|s| !s.is_empty()),
-                    start: serde_json::from_str(attributes.get("start").unwrap()).unwrap(),
-                    end: serde_json::from_str(attributes.get("end").unwrap()).unwrap(),
+                    attributes: Some(current_line.attributes).filter(|s| !s.is_empty()),
+                    start: Position::new(1, 1),
+                    end: Position::new(1, 1),
                 };
-                Block::Heading(heading)
+                Ok(Block::Heading(heading))
             }
             "Paragraph" => {
                 let paragraph = Paragraph {
                     content: self.read_inlines(current_line.content.clone()),
                 };
-                Block::Paragraph(paragraph)
+                Ok(Block::Paragraph(paragraph))
             }
             "VerbatimBlock" => {
                 let verbatim = VerbatimBlock {
                     content: current_line.content.clone(),
```

Comment:
```suggestion
                    content: current_line.content.clone(), // TODO: use inline parser, but only allow `logic` and plain text.
```

Note for me. Will change once logic will be integrated.